### PR TITLE
VB-3452 remove tap as a transitional location type

### PR DESCRIPTION
**Pull request by dhirajshettymoj**
## What does this pull request do?

Allow sessions to be added only for users in TAP locations.
## What is the intent behind these changes?

VB-3452 - new functionality.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/sessions/UpdateSessionTemplateDto.kt [hunk 1]
```diff
@@ -44,4 +44,7 @@ data class UpdateSessionTemplateDto(
 
   @Schema(description = "list of group references for allowed prisoner incentive levels", required = false)
   val incentiveLevelGroupReferences: List<String>? = null,
+
+  @Schema(description = "If a session is a TAP session.", required = true)
+  val isTAPSession: Boolean,
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/sessions/UpdateSessionTemplateDto.kt:49]: This should be in the location group,... may be actually add an exclude param instead

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/session/SessionTemplate.kt [hunk 2]
```diff
@@ -93,4 +93,7 @@ class SessionTemplate(
 
   @Column(name = "active", nullable = false)
   var active: Boolean,
+
+  @Column(name = "tapSession", nullable = false)
+  var isTapSession: Boolean,
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/model/entity/session/SessionTemplate.kt:98]: does not belong here, it's not it's concern should be in SessionLocationGroup

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt [hunk 3]
```diff
@@ -170,6 +170,10 @@ interface SessionTemplateRepository : JpaRepository<SessionTemplate, Long> {
   @Query("Update SessionTemplate s set s.active = :isActive WHERE s.reference = :reference")
   fun updateActiveByReference(reference: String, isActive: Boolean): Int
 
+  @Modifying
+  @Query("Update SessionTemplate s set s.isTapSession = :isTapSession WHERE s.reference = :reference")
+  fun updateIsTapSessionByReference(reference: String, isTapSession: Boolean): Int
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/SessionTemplateRepository.kt:175]: should be in SESSION_LOCATION_GROUP table as an exclude, this needs to be data driven

**Review by AledWEvans — CHANGES_REQUESTED** [review]
I think we need to re-think this, it should be data driven so you can add TAP in the admin page and then choose to include it or exclude it

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/sessions/CreateSessionTemplateDto.kt [hunk 4]
```diff
@@ -58,4 +58,7 @@ data class CreateSessionTemplateDto(
 
   @Schema(description = "list of group references for allowed prisoner incentive levels", required = false)
   val incentiveLevelGroupReferences: List<String>? = listOf(),
+
+  @Schema(description = "If a session is a TAP session.", required = true)
+  val isTAPSession: Boolean,
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/sessions/CreateSessionTemplateDto.kt:63]: I think it should be data driven,...

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonerService.kt [hunk 5]
```diff
@@ -53,10 +53,15 @@ class PrisonerService(
...
 
   private fun isPrisonerInTemporaryLocation(levels: List<PrisonerHousingLevelDto>): Boolean {
-    if (levels.isNotEmpty()) {
-      return TransitionalLocationTypes.contains(levels[0].code)
-    }
-    return false
+    return (levels.isNotEmpty() && TransitionalLocationTypes.contains(levels[0].code))
+  }
+
+  fun isPrisonerInTAPTemporaryLocation(prisonerHousingLocation: PrisonerHousingLocationsDto?): Boolean {
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonerService.kt:59]: is this the correct naming?

Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonerService.kt:59]: sessionTemplates = if (sessionTemplates.any { it.isTapSession } && prisonerService.isPrisonerInTAPTemporaryLocation(prisonerHousingLocation)) {
      sessionTemplates.filter { it.isTapSession }
    } else {
      sessionTemplates.filter {
        isAvailableToLocation(it, sessionTemplates, prisonerId, prisonCode, prisonerHousingLocation)
          .and(isAvailableToCategory(it, sessionTemplates, prisoner.category))
          .and(isAvailableToIncentiveLevel(it, sessionTemplates, prisoner.incentiveLevel))
      }
    }

I think all the logic should be data driven not a hard coded connection, so the location group locations should have TAP in it, set in admin

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonerService.kt [hunk 6]
```diff
@@ -52,16 +53,19 @@ class PrisonerService(
...
-
-  fun getLevelsMapForPrisoner(prisonerHousingLocationsDto: PrisonerHousingLocationsDto): Map<PrisonerHousingLevels, String?> {
-    with(prisonerHousingLocationsDto) {
-      val housingLevel = if (isPrisonerInTemporaryLocation(levels)) lastPermanentLevels else levels
+  fun getLevelsMapForPrisoner(prisonerHousingLocation: PrisonerHousingLocationsDto, sessionTemplates: List<SessionTemplate>?): Map<PrisonerHousingLevels, String?> {
+    with(prisonerHousingLocation) {
+      val housingLevel = if (isPrisonerInTemporaryLocation(levels)) {
+        val transitionalLocation = if (levels.isNotEmpty()) levels[0].code else ""
+        // if there are sessions for the prisoners temporary location - level one code needs to be that transitional location
+        if (sessionTemplates != null && hasPrisonGotSessionsWithPrisonersTransitionalLocation(sessionTemplates, transitionalLocation)) {
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonerService.kt:61]: if prisoner is in temporary location and there are session templates for that location in the prison show only those sessions.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/sessions/CreateSessionTemplateDto.kt (hunk 4 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/sessions/CreateSessionTemplateDto.kt]: Thanks. Yes, as you suggested it is now data driven with the includeLocationGroups flag.

Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/sessions/CreateSessionTemplateDto.kt]: Updated to be data driven as suggested with the includeLocationGroups flag.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonerService.kt (hunk 5 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonerService.kt]: Updated to be data driven as suggested with the includeLocationGroups flag.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonerService.kt [hunk 7]
```diff
@@ -52,16 +53,19 @@ class PrisonerService(
...
-    return false
-  }
-
-  fun getLevelsMapForPrisoner(prisonerHousingLocationsDto: PrisonerHousingLocationsDto): Map<PrisonerHousingLevels, String?> {
-    with(prisonerHousingLocationsDto) {
-      val housingLevel = if (isPrisonerInTemporaryLocation(levels)) lastPermanentLevels else levels
+  fun getLevelsMapForPrisoner(prisonerHousingLocation: PrisonerHousingLocationsDto, sessionTemplates: List<SessionTemplate>?): Map<PrisonerHousingLevels, String?> {
+    with(prisonerHousingLocation) {
+      val housingLevel = if (isPrisonerInTemporaryLocation(levels)) {
+        val transitionalLocation = if (levels.isNotEmpty()) levels[0].code else ""
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonerService.kt:59]: levels.firstOrNull()?:""

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonerService.kt [hunk 8]
```diff
@@ -74,6 +82,13 @@ class PrisonerService(
...
 
+  fun getHousingLevelForPrisonerInTemporaryLocation(prisonersTransitionalLocation: String): List<PrisonerHousingLevelDto> {
+    val prisonerHousingLocationForTransitionalPrisoners = mutableListOf<PrisonerHousingLevelDto>()
+    val prisonerHousingLocation = PrisonerHousingLevelDto(level = LEVEL_ONE.level, code = prisonersTransitionalLocation, description = prisonersTransitionalLocation)
+    prisonerHousingLocationForTransitionalPrisoners.add(prisonerHousingLocation)
+    return prisonerHousingLocationForTransitionalPrisoners.toList()
+  }
+
   private fun getHousingLevelByLevelNumber(levels: List<PrisonerHousingLevelDto>, housingLevel: PrisonerHousingLevels): PrisonerHousingLevelDto? {
     return levels.stream().filter { level -> level.level == housingLevel.level }.findFirst().orElse(null)
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonerService.kt:93]: levels.asSequence()

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonerService.kt [hunk 9]
```diff
@@ -74,6 +82,13 @@ class PrisonerService(
     }
   }
 
+  fun getHousingLevelForPrisonerInTemporaryLocation(prisonersTransitionalLocation: String): List<PrisonerHousingLevelDto> {
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonerService.kt:85]: private

**Pull request merged**